### ISSUE TITLE
[Bug]: mpl.colors.CenteredNorm has wrong fig.colorbar in some situation

### ISSUE THREAD

**Issue by ickc** [matplotlib/matplotlib#21630]
### Bug summary

In some specific situation, when `fig.colorbar` is used with `mpl.colors.CenteredNorm`, the colorbar has wrong color gradient.

### Code for reproduction

This example is the bug:

```python
import matplotlib as mpl
import matplotlib.pyplot as plt

cmap = "coolwarm"
halfrange = 1.6012970450803282e-06
norm = mpl.colors.CenteredNorm(0, halfrange)
fig, axes = plt.subplots(2)
for ax in axes:
    im = ax.imshow(
        [
            [-1.543064702687502e-05, 0., 1.3409952347604702e-05],
            [-1.5816019543010574e-06, 0, 1.6012970450803282e-06],
            [0., 0. ,0.],
        ],
        cmap=cmap,
        norm=norm,
    )
    fig.colorbar(im, ax=ax)
fig.savefig("bug-CenteredNorm.png")
```

And bug disappeared when changing `CenteredNorm` to `TwoSlopeNorm`:

```python
import matplotlib as mpl
import matplotlib.pyplot as plt

cmap = "coolwarm"
halfrange = 1.6012970450803282e-06
norm = mpl.colors.TwoSlopeNorm(0, -halfrange, halfrange)
fig, axes = plt.subplots(2)
for ax in axes:
    im = ax.imshow(
        [
            [-1.543064702687502e-05, 0., 1.3409952347604702e-05],
            [-1.5816019543010574e-06, 0, 1.6012970450803282e-06],
            [0., 0. ,0.],
        ],
        cmap=cmap,
        norm=norm,
    )
    fig.colorbar(im, ax=ax)
fig.savefig("bug-TwoSlopeNorm.png")
```

### Actual outcome

![bug-CenteredNorm](https://user-images.githubusercontent.com/10970208/141667042-0d75ddef-0e2b-4234-8319-7ef14e4ee87a.png)

### Expected outcome

![bug-TwoSlopeNorm](https://user-images.githubusercontent.com/10970208/141667047-5a802b77-3045-4a70-8dda-1eef2772d194.png)

### Operating system

macOS 11.6 20G165 x86_64 

### Matplotlib Version

    matplotlib                3.4.3            py39h6e9494a_2    conda-forge
    matplotlib-base           3.4.3            py39hb07454d_2    conda-forge

### Matplotlib Backend

`module://ipykernel.pylab.backend_inline`

### Python version

```
python                    3.9.5                h88f2d9e_3  

Python 3.9.5 (default, May 18 2021, 12:31:01)
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
```

### Jupyter version

```
jupyterhub                1.4.2            py39h6e9494a_0    conda-forge
jupyterhub-base           1.4.2            py39h6e9494a_0    conda-forge
jupyterlab                3.1.14             pyhd8ed1ab_0    conda-forge
```

### Installation

conda

### Conda channel

conda-forge

**Comment by greglucas** [comment]
This appears to be fixed on the upcoming v3.5 release (you can install the release candidate now with `pip install matplotlib==3.5.0rc1`).

**Comment by timhoffm** [comment]
I can confirm that this is fixed in v3.5.
